### PR TITLE
Added preferences file, where will be saved data about field

### PR DESCRIPTION
**Pull request by ChippZanuff**

### CODE REVIEW

File: app/src/main/java/com/example/tes/sapper/Preferences.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        if(this.settings.exists())
+        {
+            this.cellsAmount = this.preferences.getInt(this.CELLS_AMOUNT, this.getCellsAmount());
+            this.minesAmount = this.preferences.getInt(this.MINES_AMOUNT, this.getMinesAmount());
+            this.numRows = this.preferences.getInt(this.NUM_ROWS, this.getNumRows());
+            this.numCols = this.preferences.getInt(this.NUM_COLS, this.getNumCols());
+        }
+        else
+        {
+            this.createPreferencesFile();
```
Comment (ChippZanuff) [inline app/src/main/java/com/example/tes/sapper/Preferences.java]: .

File: app/src/main/java/com/example/tes/sapper/Preferences.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        String settingsFilePath = this.context.getApplicationInfo().dataDir + "/shared_prefs/" + SETTINGS_FILE + ".xml";
+        this.settings = new File(settingsFilePath);
+
+        if(this.settings.delete());
+
+        this.getPreferencesFileValues();
+
+    }
+
+    private void createPreferencesFile()
```
Comment (ChippZanuff) [inline app/src/main/java/com/example/tes/sapper/Preferences.java]: .

File: app/src/main/java/com/example/tes/sapper/Preferences.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        this.log = log;
+        this.log.setTAG(getClass().getSimpleName());
+        this.context = context;
+
+        String SETTINGS_FILE = "Settings";
+        this.preferences = this.context.getSharedPreferences(SETTINGS_FILE, Context.MODE_PRIVATE);
+
+        String settingsFilePath = this.context.getApplicationInfo().dataDir + "/shared_prefs/" + SETTINGS_FILE + ".xml";
+        this.settings = new File(settingsFilePath);
```
Comment (ChippZanuff) [inline app/src/main/java/com/example/tes/sapper/Preferences.java]: .

File: app/src/main/java/com/example/tes/sapper/Preferences.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+        if(this.settings.delete());
```
Comment (ChippZanuff) [inline app/src/main/java/com/example/tes/sapper/Preferences.java]: .

File: app/src/main/java/com/example/tes/sapper/ImageAdapter.java [hunk 5]
```diff
@@ -12,12 +12,14 @@
...
     private Logger log;
+    private Preferences preferences;
     //int[] thumbIds;
 
-    public ImageAdapter(Context context, int numOfCells, Logger log)
+    public ImageAdapter(Context context, Preferences preferences, Logger log)
     {
         this.context = context;
-        countOfCell = numOfCells;
+        this.countOfCell = preferences.getCellsAmount();
```
Comment (alexey-fut) [inline app/src/main/java/com/example/tes/sapper/ImageAdapter.java]: Same as first..

File: app/src/main/java/com/example/tes/sapper/Board.java [hunk 6]
```diff
@@ -4,32 +4,30 @@
...
-        this.rows = rows;
         this.cells = this.fieldCreate();
         this.setMineField();
     }
 
     private void setMineField()
     {
-        int minesAmount = this.amountOfCells / this.amountOfMines;
-        int availableRow = this.amountOfCells / this.numColumns;
+        int minesAmount = this.preferences.getCellsAmount() / this.preferences.getMinesAmount();
```
Comment (alexey-fut) [inline app/src/main/java/com/example/tes/sapper/Board.java]: encapsulate it

File: app/src/main/java/com/example/tes/sapper/Board.java [hunk 7]
```diff
@@ -4,32 +4,30 @@
...
+        int availableRow = this.preferences.getCellsAmount() / this.preferences.getNumCols();
```
Comment (alexey-fut) [inline app/src/main/java/com/example/tes/sapper/Board.java]: encapsulate it

File: app/src/main/java/com/example/tes/sapper/Board.java [hunk 8]
```diff
@@ -4,32 +4,30 @@
...
         CellParam cell;
 
         while(true)
         {
             int mineRow = (int) (Math.random() * (availableRow));
-            int mineCol = (int) (Math.random() * (this.numColumns));
+            int mineCol = (int) (Math.random() * (this.preferences.getNumCols()));
```
Comment (alexey-fut) [inline app/src/main/java/com/example/tes/sapper/Board.java]: encapsulate it

File: app/src/main/java/com/example/tes/sapper/Board.java [hunk 9]
```diff
@@ -4,32 +4,30 @@
...
     {
-        int minesAmount = this.amountOfCells / this.amountOfMines;
-        int availableRow = this.amountOfCells / this.numColumns;
+        int minesAmount = this.preferences.getCellsAmount() / this.preferences.getMinesAmount();
+        int availableRow = this.preferences.getCellsAmount() / this.preferences.getNumCols();
         CellParam cell;
 
         while(true)
         {
             int mineRow = (int) (Math.random() * (availableRow));
```
Comment (alexey-fut) [inline app/src/main/java/com/example/tes/sapper/Board.java]: encapsulate it

File: app/src/main/java/com/example/tes/sapper/Board.java [hunk 10]
```diff
@@ -50,11 +48,11 @@ private void setMineField()
         ArrayList<ArrayList<CellParam>> field = new ArrayList<>();
 
 
-        for(int row = 0; row < this.amountOfCells / this.numColumns; row++)
+        for(int row = 0; row < this.preferences.getCellsAmount() / this.preferences.getNumCols(); row++)
```
Comment (alexey-fut) [inline app/src/main/java/com/example/tes/sapper/Board.java]: encapsulate it

File: app/src/main/java/com/example/tes/sapper/Board.java [hunk 11]
```diff
@@ -4,32 +4,28 @@
...
     {
-        int minesAmount = this.amountOfCells / this.amountOfMines;
-        int availableRow = this.amountOfCells / this.numColumns;
         CellParam cell;
 
         while(true)
         {
-            int mineRow = (int) (Math.random() * (availableRow));
-            int mineCol = (int) (Math.random() * (this.numColumns));
+            int mineRow = (int) (Math.random() * (this.preferences.getNumRows()));
```
Comment (alexey-fut) [inline app/src/main/java/com/example/tes/sapper/Board.java]: ?

File: app/src/main/java/com/example/tes/sapper/Board.java [hunk 12]
```diff
@@ -48,30 +44,31 @@ private void setMineField()
     private ArrayList<ArrayList<CellParam>> fieldCreate()
     {
         ArrayList<ArrayList<CellParam>> field = new ArrayList<>();
+        int count = 0;
 
-
-        for(int row = 0; row < this.amountOfCells / this.numColumns; row++)
+        while(count < this.preferences.getCellsAmount())
```
Comment (alexey-fut) [inline app/src/main/java/com/example/tes/sapper/Board.java]: why 'for' was replaced by while?

File: app/src/main/java/com/example/tes/sapper/Preferences.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    private int numRows;
+
+    public Preferences(SharedPreferences preferences, File settingsFile, Logger log)
+    {
+        this.log = log;
+        this.log.setTAG(getClass().getSimpleName());
+
+        this.settingsFile = settingsFile;
+        this.preferences = preferences;
+        this.numRows = this.cellsAmount / this.numCols;
```
Comment (alexey-fut) [inline app/src/main/java/com/example/tes/sapper/Preferences.java]: ?

File: app/src/main/java/com/example/tes/sapper/Board.java [hunk 14]
```diff
@@ -38,7 +34,7 @@ private void setMineField()
                 this.flagsLeft++;
             }
 
-            if(this.flagsLeft == minesAmount)
+            if(this.flagsLeft == this.preferences.getMineCounter())
```
Comment (ChippZanuff) [inline app/src/main/java/com/example/tes/sapper/Board.java]: .

File: app/src/main/java/com/example/tes/sapper/Board.java (hunk 12 above)
Comment (ChippZanuff) [inline app/src/main/java/com/example/tes/sapper/Board.java]: .

File: app/src/main/java/com/example/tes/sapper/Board.java [hunk 15]
```diff
@@ -100,17 +97,17 @@ public CellParam getCellById(int row ,int col)
 
     public int getRows()
```
Comment (ChippZanuff) [inline app/src/main/java/com/example/tes/sapper/Board.java]: .

File: app/src/main/java/com/example/tes/sapper/Preferences.java (hunk 13 above)
Comment (ChippZanuff) [inline app/src/main/java/com/example/tes/sapper/Preferences.java]: .

File: app/src/main/java/com/example/tes/sapper/Board.java [hunk 16]
```diff
@@ -38,7 +35,7 @@ private void setMineField()
                 this.flagsLeft++;
             }
 
-            if(this.flagsLeft == minesAmount)
+            if(this.flagsLeft == this.preferences.getMineCounter())
```
Comment (alexey-fut) [inline app/src/main/java/com/example/tes/sapper/Board.java]: better to check as greater or equal

File: app/src/main/java/com/example/tes/sapper/Preferences.java [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    public void setPreferencesCellsAmount(int value)
+    {
+        this.setPreferencesValue("Cells Amount", value);
+    }
+
+    public void setPreferencesMinesAmount(int value)
+    {
+        this.setPreferencesValue("Mines Amount", value);
```
Comment (alexey-fut) [inline app/src/main/java/com/example/tes/sapper/Preferences.java]: Hardcoded value should be moved to const

**Pull request merged**